### PR TITLE
Add more type hints

### PR DESCRIPTION
**Pull request by Illviljan**
Add more type hints.

### CODE REVIEW

File: flox/xarray.py [hunk 1]
```diff
@@ -276,7 +276,7 @@ def xarray_reduce(
     exclude_dims = tuple(d for d in ds.dims if d not in grouper_dims and d not in dim_tuple)
 
     try:
-        xr.align(ds, *by_da, join="exact")
+        xr.align(ds, *by_da, join="exact", copy=False)
```
Comment (dcherian) [inline flox/xarray.py:279]: Nice find! Would be nice to open an issue  at xarray asking for `xarray.testing.assert_aligned` or something similar.

File: flox/core.py [hunk 2]
```diff
@@ -1550,7 +1544,10 @@ def _factorize_multiple(by, expected_groups, any_by_dask, reindex):
             *by_,
             chunks=tuple(chunks.values()),
             meta=np.array((), dtype=np.int64),
-            **kwargs,
```
Comment (dcherian) [inline flox/core.py:1553]: again, I prefer the dict  version

Comment (Illviljan) [reply flox/core.py:1553]: Using kwargs breaks type checking. Everything becomes Any then and then you are forced to use isinstance checks.

**Comment by Illviljan** [comment]
```
flox/core.py:1615: error: Argument 1 to "len" has incompatible type "Optional[Any]"; expected "Sized"  [arg-type]
Generated Cobertura report: /home/runner/work/flox/flox/mypy_report/cobertura.xml
Installing missing stub packages:
flox/core.py:1683: error: "Unpack" support is experimental, use --enable-incomplete-feature=Unpack to enable  [misc]
/home/runner/micromamba-root/envs/flox-tests/bin/python -m pip install types-PyYAML types-colorama types-psutil types-xxhash
flox/core.py:1950: error: Incompatible return value type (got "Tuple[Any, ...]", expected "Tuple[Any, Any]")  [return-value]


Generated Cobertura report: /home/runner/work/flox/flox/mypy_report/cobertura.xml
Found 3 errors in 1 file (checked 11 source files)
```



@dcherian, this error:
```
flox/core.py:1615: error: Argument 1 to "len" has incompatible type "Optional[Any]"; expected "Sized"  [arg-type]
```
comes from
https://github.com/xarray-contrib/flox/blob/cbd52efe83d4f3ae23fb9d2bcc16ad718f740472/flox/core.py#L1555-L1560
where `len(f)` can be `len(None)` which will fail. Why is None set for dask arrays?

**Comment by dcherian** [comment]
Oh we can't group by multiple dask arrays and have no expected_groups provided.

So that code is supposed to loop through all `by` variables together with `expected_groups`.

If the expected group labels are not provided, we detect them automatically for numpy arrays (this actually seems not great now but might have to do with it being hard to return multiple arrays from `_lazy_factorize_wrapper`).

The mypy error is right: we should explicitly raise an error when `e is None and f is None`

File: flox/core.py [hunk 3]
```diff
@@ -1550,25 +1603,53 @@ def _factorize_multiple(by, expected_groups, any_by_dask, reindex):
...
+        )
 
-    final_groups = tuple(
-        found if expect is None else expect.to_numpy()
-        for found, expect in zip(found_groups, expected_groups)
-    )
+    # final_groups = tuple(
+    #     found if expect is None else expect.to_numpy()
+    #     for found, expect in zip(found_groups, expected_groups)
+    # )
```
Comment (Illviljan) [inline flox/core.py:1648]: As I understand it `found_groups` from `factorize_` is just a better version of `expected_groups`. 
So I don't see why `expected_groups` in `final_groups` is prioritized. 

Seems to pass the tests skipping this loop at least.

**Comment by Illviljan** [comment]
Starting to get quite a bit of changes. Think I'll stop here.

File: flox/core.py (hunk 3 above)
Comment (dcherian) [reply flox/core.py]: It's because `set(expected_groups) < set(found_groups)`  possibly.

File: flox/core.py [hunk 4]
```diff
@@ -665,38 +731,42 @@ def chunk_reduce(
...
     # avoid by factorizing again so indices=[2,2,2] is changed to
     # indices=[0,0,0]. This is necessary when combining block results
     # factorize can handle strings etc unlike digitize
-    group_idx, groups, found_groups_shape, _, size, props = factorize_(
-        (by,), axis, expected_groups=(expected_groups,), reindex=reindex, sort=sort
+    group_idx, grps, found_groups_shape, _, size, props = factorize_(
+        (by,), axes, expected_groups=(expected_groups,), reindex=reindex, sort=sort
     )
-    groups = groups[0]
+    groups = grps[0]
```
Comment (dcherian) [inline flox/core.py:764]:
```suggestion
    groups, = grps
```

Comment (dcherian) [inline flox/core.py:764]:
```suggestion
```

File: flox/core.py [hunk 5]
```diff
@@ -1550,25 +1616,39 @@ def _factorize_multiple(by, expected_groups, any_by_dask, reindex):
...
         )
-    else:
-        group_idx, found_groups, grp_shape = factorize_(by, **kwargs)
 
-    final_groups = tuple(
-        found if expect is None else expect.to_numpy()
-        for found, expect in zip(found_groups, expected_groups)
-    )
+        fg, gs = [], []
+        for b, e in zip(by, expected_groups):
```
Comment (dcherian) [inline flox/core.py:1626]:
```suggestion
        for by_, expect in zip(by, expected_groups):
```

File: flox/core.py [hunk 6]
```diff
@@ -1550,25 +1616,39 @@ def _factorize_multiple(by, expected_groups, any_by_dask, reindex):
...
+
+                f = pd.unique(b.reshape(-1))
+            else:
+                f = e.to_numpy()
+
+            fg.append(f)
+            gs.append(len(f))
+
+        found_groups = tuple(fg)
+        grp_shape = tuple(gs)
```
Comment (dcherian) [inline flox/core.py:1641]: I prefer the previous comprehension.

File: flox/core.py (hunk 3 above)
Comment (dcherian) [reply flox/core.py]: Sounds like we need a extra test.

File: flox/core.py [hunk 7]
```diff
@@ -665,38 +731,42 @@ def chunk_reduce(
...
+        nax = 1
 
     # if indices=[2,2,2], npg assumes groups are (0, 1, 2);
     # and will return a result that is bigger than necessary
     # avoid by factorizing again so indices=[2,2,2] is changed to
     # indices=[0,0,0]. This is necessary when combining block results
     # factorize can handle strings etc unlike digitize
-    group_idx, groups, found_groups_shape, _, size, props = factorize_(
-        (by,), axis, expected_groups=(expected_groups,), reindex=reindex, sort=sort
+    group_idx, grps, found_groups_shape, _, size, props = factorize_(
```
[the remark this answers is not in the document]
Comment (Illviljan) [reply flox/core.py:761]: Have to do something like this: `(groups, *_) = (1,2,3,4)`. Is that nicer than `groups = grps[0]`?

File: flox/core.py (hunk 5 above)
Comment (Illviljan) [reply flox/core.py]: Done.

File: flox/core.py (hunk 6 above)
Comment (Illviljan) [reply flox/core.py:1641]: Previous comprehension didn't have the same typing for `found_groups` as `factorize_` and `grp_shape` could crash (`len(None)`).

**Pull request merged**